src/glib/mine/clustering.h

```diff
@@ -729,16 +752,66 @@ void TAbsKMeans<TCentroidType>::InitCentroids(TVec<TIntFltKdV>& CentroidVV, cons
 }
 
 template<class TCentroidType>
+void TAbsKMeans<TCentroidType>::InitCentroids(TFltVV& CentroidVV, const TFltVV& FtrVV) {
+    const int Rows = GetDataDim(FtrVV);
+    const int Cols = GetDataCount(FtrVV);
+    // construct the centroid matrix
+    CentroidVV.Gen(Rows, Cols);
+    for (int ColN = 0; ColN < Cols; ColN++) {
+        for (int RowN = 0; RowN < Rows; RowN++) {
+            CentroidVV.PutXY(RowN, ColN, FtrVV(RowN, ColN));
+        }
+    }
+}
+
+template<class TCentroidType>
+void TAbsKMeans<TCentroidType>::InitCentroids(TFltVV& CentroidVV, const TVec<TIntFltKdV>& FtrVV) {
+    const int Rows = GetDataDim(FtrVV);
+    const int Cols = GetDataCount(FtrVV);
+    // construct the centroid matrix
+    CentroidVV.Gen(Rows, Cols);
+    for (int ColN = 0; ColN < Cols; ColN++) {
+        const int Els = FtrVV[ColN].Len();
+        for (int ElN = 0; ElN < Els; ElN++) {
+            CentroidVV.PutXY(FtrVV[ColN][ElN].Key, ColN, FtrVV[ColN][ElN].Dat);
+        }
+    }
+}
+
+template<class TCentroidType>
+void TAbsKMeans<TCentroidType>::InitCentroids(TVec<TIntFltKdV>& CentroidVV, const TFltVV& FtrVV) {
+    const int Rows = GetDataDim(FtrVV);
+    const int Cols = GetDataCount(FtrVV);
+    // construct the centroid matrix
+    CentroidVV.Gen(Cols);
+    for (int ColN = 0; ColN < Cols; ColN++) {
+        for (int RowN = 0; RowN < Rows; RowN++) {
+            if (FtrVV(RowN, ColN) != 0.0) { CentroidVV[ColN].Add(TIntFltKd(RowN, FtrVV(RowN, ColN))); }
+        }
+    }
+}
+
+template<class TCentroidType>
+void TAbsKMeans<TCentroidType>::InitCentroids(TVec<TIntFltKdV>& CentroidVV, const TVec<TIntFltKdV>& FtrVV) {
+    const int Cols = GetDataCount(FtrVV);
+    // construct the centroid matrix
+    CentroidVV.Gen(Cols);
+    for (int ColN = 0; ColN < Cols; ColN++) {
+        CentroidVV[ColN] = FtrVV[ColN];
+    }
+}
+
```

Comment:
These four init methods should reuse code from linalg.h (type transformations) and copy assignment operators.